rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -0,0 +1,978 @@
+use blockifier::execution::contract_class::{
+    ContractClass as BlockifierContractClass, ContractClassV0, ContractClassV0Inner,
+};
+use cairo_lang_starknet::casm_contract_class::CasmContractClass;
+use cairo_lang_starknet::contract_class::{
+    ContractClass as SierraContractClass, ContractEntryPoints,
+};
+use cairo_vm::types::program::Program;
+use cairo_vm::vm::runners::cairo_runner::ExecutionResources as VmExecutionResources;
+use core::fmt;
+use dotenv::dotenv;
+use serde::{Deserialize, Deserializer};
+use serde_json::json;
+use serde_with::{serde_as, DeserializeAs};
+use starknet::core::types::ContractClass as SNContractClass;
+use starknet_api::block::{BlockNumber, BlockTimestamp};
+use starknet_api::core::{ChainId, ClassHash, EntryPointSelector};
+use starknet_api::deprecated_contract_class::EntryPointOffset;
+use starknet_api::hash::StarkFelt;
+use starknet_api::serde_utils::PrefixedBytesAsHex;
+use starknet_api::transaction::{InvokeTransaction, Transaction, TransactionHash};
+use starknet_api::{core::ContractAddress, hash::StarkHash, state::StorageKey};
+use std::collections::HashMap;
+use std::env;
+use std::sync::Arc;
+use thiserror::Error;
+
+/// Starknet chains supported in Infura.
+#[derive(Debug, Clone, Copy)]
+pub enum RpcChain {
+    MainNet,
+    TestNet,
+    TestNet2,
+}
+
+impl fmt::Display for RpcChain {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RpcChain::MainNet => write!(f, "starknet-mainnet"),
+            RpcChain::TestNet => write!(f, "starknet-goerli"),
+            RpcChain::TestNet2 => write!(f, "starknet-goerli2"),
+        }
+    }
+}
+
+/// A [StateReader] that holds all the data in memory.
+///
+/// This implementation is uses HTTP requests to call the RPC endpoint,
+/// using Infura.
+/// In order to use it an Infura API key is necessary.
+pub struct RpcState {
+    /// Enum with one of the supported Infura chains/
+    chain: RpcChain,
+    /// Infura API key.
+    api_key: String,
+    /// Struct that holds information on the block where we are going to use to read the state.
+    block: BlockValue,
+}
+
+#[derive(Debug, Error)]
+enum RpcError {
+    #[error("RPC call failed with error: {0}")]
+    RpcCall(String),
+    #[error("Request failed with error: {0}")]
+    Request(String),
+    #[error("Failed to cast from: {0} to: {1} with error: {2}")]
+    Cast(String, String, String),
+}
+
+/// [`BlockValue`] is an Enum that represent which block we are going to use to retrieve information.
+#[allow(dead_code)]
+pub enum BlockValue {
+    /// String one of: ["latest", "pending"]
+    Tag(serde_json::Value),
+    /// Integer
+    Number(serde_json::Value),
+    /// String with format: 0x{felt252}
+    Hash(serde_json::Value),
```

Comment:
Why `serde_json::Value`? Why not an enum for `Tag`, an actual number for `Number` and a bigint (or bytes) for `Hash`?

Reply:
Hmm not really sure about that, thanks for this suggestion, it really improved this type.